core/src/main/java/org/elasticsearch/action/ActionListener.java

```diff
@@ -70,6 +71,50 @@ public void onFailure(Exception e) {
     }
 
     /**
+     * Creates a listener that listens for a response (or failure) and executes the
+     * corresponding runnable when the response (or failure) is received.
+     *
+     * @param runnable the runnable that will be called in event of success or failure
+     * @param <Response> the type of the response
+     * @return a listener that listens for responses and invokes the runnable when received
+     */
+    static <Response> ActionListener<Response> wrap(Runnable runnable) {
+        return new ActionListener<Response>() {
+            @Override
+            public void onResponse(Response response) {
+                runnable.run();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                runnable.run();
+            }
+        };
+    }
+
+    /**
+     * Converts a listener to a {@link BiConsumer} for compatibility with the {@link java.util.concurrent.CompletableFuture}
+     * api.
+     *
+     * @param listener that will be wrapped
+     * @param <Response> the type of the response
+     * @return a bi consumer that will complete the wrapped listener
+     */
+    static <Response> BiConsumer<Response, Throwable> toBiConsumer(ActionListener<Response> listener) {
+        return (tcpChannel, throwable) -> {
+            if (throwable == null) {
+                listener.onResponse(tcpChannel);
+            } else {
+                if (throwable instanceof Exception) {
+                    listener.onFailure((Exception) throwable);
+                } else {
```

Comment:
I don't think we should hide the non-exception. this will be an error and in this case we need to rethrow it? @jasontedor WDYT?